xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,256 @@
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+/*
+ListNET is adapted from the paper:
+Cao, Zhe, et al. "Learning to rank: from pairwise approach to listwise approach."
+Proceedings of the 24th international conference on Machine learning. ACM, 2007.
+*/
+
+#include "xapian-letor/ranker.h"
+#include "debuglog.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+using namespace Xapian;
+
+/// A vector holding ground truth and prediction score probability distribution vectors.
+typedef vector <vector <double> > prob_distrib_vector;
+
+ListNETRanker::~ListNETRanker() {
+    LOGCALL_DTOR(API, "ListNETRanker");
+}
+
+static double
+calculateInnerProduct(vector<double> parameters, vector<double> fvals) {
+    LOGCALL_STATIC_VOID(API, "calculateInnerProduct", parameters | fvals);
+
+    double inner_product = 0.0;
+
+    for (int i = 0; i < int(fvals.size()); ++i){
+	inner_product += parameters[i] * fvals[i];
+    }
+
+    return inner_product;
+}
+
+// From Theorem (8) in Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static prob_distrib_vector
+initializeProbability(vector<FeatureVector> feature_vectors, vector<double> & new_parameters) {
+    LOGCALL_STATIC_VOID(API, "initializeProbability", feature_vectors | new_parameters);
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+	expsum_y += exp(feature_vectors[i].get_label());
+	expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+	prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+	prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+    return prob;
+}
+
+// Equation (6) in paper Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, prob_distrib_vector prob) {
```

Comment:
Both are these are `std::vector`, so you really want to pass them by const reference not value - currently this will have to copy both the passed vectors.